src/analysis/raster/qgsrastercalcnode.cpp

```diff
@@ -220,14 +225,22 @@ bool QgsRasterCalcNode::calculate( QMap<QString, QgsRasterBlock * > &rasterData,
     QVector <QgsRasterMatrix *> matrixContainer;
     for ( int i = 0; i < mFunctionArgs.size(); ++i )
     {
-      std::unique_ptr< QgsRasterMatrix > singleMatrix( new QgsRasterMatrix( result.nColumns(), result.nRows(), nullptr, result.nodataValue() ) );
+      QgsRasterMatrix *singleMatrix = new QgsRasterMatrix( result.nColumns(), result.nRows(), nullptr, result.nodataValue() );
```

Comment:
I may have misread the code but removing unique_ptr for a raw pointer seems a step backward to me.

Reply:
@lbartoletti what's your motivation?

Reply:
I would have used std::make_unique and remove the `new` alltogether.

Reply:
Because unique_ptr is an exclusive resource and cannot be assigned, but here it is actually necessary to pass the pointer to the function "evaluateFunction". The original code took the pointer out and did not release it, resulting in a memory leak, so the naked pointer is used.

Reply:
In that case you could make matrixContainer  a vector of unique_ptr and sd::move it

Reply:
Actually I'm curious why shared_ptr is not Ok, then I will change the parameter of evaluateFunction to a vector of unique_ptr. Actually I'm curious why shared_ptr is not used, I don't see it in the whole analysis module.used, I don't see it in the whole analysis module.

Reply:
Ok, then I will change the parameter of evaluateFunction to a vector of unique_ptr. Actually I'm curious why shared_ptr is not used, I don't see it in the whole analysis module.

Reply:
Most part of the analysis code was probably written many years ago. 

Shared pointers are generally fine, but in this particular case a unique pointer is probably a better fit.

Reply:
OK, got it. I made a modification to use unique_ptr.https://github.com/qgis/QGIS/pull/53416/commits/7f360cbde844d0197cd6672748eefacdd7b1669b